aea/protocols/base.py

```diff
@@ -48,7 +49,7 @@ class Message:
     protocol_specification_id = None  # type: PublicId
     serializer = None  # type: Type["Serializer"]
 
-    __slots__ = ("_slots", "_to", "_sender")
+    __slots__ = ("_slots", "_to", "_sender", "_envelope_context")
```

Comment:
We need to expose the context on the message, otherwise the skill is unaware which route a message took.

Reply:
not sure it's a good idea, it's obvious that `Handler.handle` accepts Message only instances.
but EnvelopeContext is out of the Messages/Protocols scope.

suppose would be nice to make Handler.handle to be able to accept some extra parameters with message related details. This will brake current api, but! it's possible to make it smart, and pass extra details if  handlers support it

Reply:
I thought about it. But it doesn't really help. Because then you have to in the skill write a lot of custom code to pass the envelope context around. The envelope context is really just a general message context. So we can simply rename it and use it here.

Reply:
Maybe we can also simplify the envelope context to be just a string? That would be less overhead.

Reply:
Another idea is to have a `MessageHeader`/`MessageContext` class, with all the attributes we want to expose (and nothing else!) to the user code. It also helps to avoid conflating two concepts (Envelope != Message), as @solarw pointed out. And could serve in the future if we want to put more information only in the message header and not in the envelope context.

Reply:
Resolved now on the multiplexer side to avoid message memory bloat and for KISS reasons.